[PR] Adding `WYSIWYG` initial implementation

LuchoTurtle opened the pull request:
closes #1 

This adds an implementation example using https://github.com/singerdmx/flutter-quill and documents the options available in order to fulfill the requirements stated in #1.

LuchoTurtle commented:
Kinda frustrating having to manually bump the `SDK` to version 21 because [`flutter_quill_extensions`](https://github.com/singerdmx/flutter-quill/tree/master/flutter_quill_extensions) is using a `gallery_saver` library that is tossing the error:

```
Launching lib/main.dart on ONEPLUS A6010 in debug mode...
main.dart:1
/Users/lucho/Documents/dwyl/flutter-wysiwyg-editor-tutorial/android/app/src/debug/AndroidManifest.xml Error:
	uses-sdk:minSdkVersion 19 cannot be smaller than version 21 declared in library [:gallery_saver] /Users/lucho/Documents/dwyl/flutter-wysiwyg-editor-tutorial/build/gallery_saver/intermediates/merged_manifest/debug/AndroidManifest.xml as the library might be using APIs not available in 19
	Suggestion: use a compatible library with a minSdk of at most 19,
		or increase this project's minSdk version to at least 21,
		or use tools:overrideLibrary="carnegietechnologies.gallery_saver" to force usage (may lead to runtime failures)
```

They should either upgrade or use another library, as stated in https://github.com/singerdmx/flutter-quill/issues/1350 :\

nelsonic commented:
That _is_ frustrating. 🤦‍♂️ 
Are you able to `Fork` it and make the necessary updates then create a PR to update upstream repo?

LuchoTurtle commented:
Perhaps but I think I can work with having `minSdkVersion` set to `21` (I just needed to upgrade from `20` to `21`, so it's not THAT big of a deal.

I'll see to it after sorting through this epic.

LuchoTurtle commented:
Been trying to paste images directly into the clipboard and not from the gallery (as an optional route) but it doesn't work in Android devices (solely, it works on other platforms).
This is because of this line:

https://github.com/singerdmx/flutter-quill/blob/5dca6008cde707beb46291b6fbf71af6dbbf7bfb/lib/src/widgets/raw_editor.dart#L1534

Unfortunately, this is [not straightforward in Android devices](https://stackoverflow.com/questions/65432586/clipboard-an-image-in-flutter). I'll keep doing the "opening gallery and choosing" route.

LuchoTurtle commented:
While this works on mobile, it does not work on `web`. I've explained one route that might get to work in https://github.com/singerdmx/flutter-quill/issues/976. 🙆‍♂️

nelsonic commented:
@LuchoTurtle please consider opening a _separate_ issue for pasting images from clipboard for `Android` so that you can ship the basic version of this. I know you want to get _everything_ on the requirements working. But it's better to ship something _basic_ than get stuck on a "nice to have" feature like this. 👌 

Saw your comment on using `universal_io`. Seems like a no-brainer to use it. 💭

LuchoTurtle commented:
Opened a PR that addresses an issue that I had embedding an image  -> https://github.com/singerdmx/flutter-quill/pull/1367.

Although it works, I'm testing this with an image URL, not a local file. I don't believe this is actually possible, so we'd need to upload an image to our `S3` bucket.

LuchoTurtle commented:
I've dismissed the previous PR because it doesn't properly work on mobile devices.

Ugh, having this working on both `web` and `mobile` is utterly frustrating.

nelsonic commented:
@LuchoTurtle please include a _lot_ more detail about your frustration. 🙏 
If something isn't working on a particular platform e.g. `Web` 🕸️ 
Please just open an issue and include screenshots, `GIFs` or `Video` explaining the issue. 🖼️ 
It might be that the insight you provide helps us to consider an alternative approach. 💭

LuchoTurtle commented:
I've added more detail and following steps to the comment 👌

LuchoTurtle commented:
I'm still in the process of writing the README detailing the steps needed to get everything working (even with the caveat that images on the web do not work as of now, they are **correctly embeded with a given image URL**, we just need to have it upload to a third-party service like our `imgup`).

Testing image uploads in `testWidgets`, as detailed in previous comments and in https://github.com/singerdmx/flutter-quill/issues/1389 have not been successful. 
I've tried c[hannel mocking](https://docs.flutter.dev/testing/plugins-in-tests#mock-the-platform-channel) (even though it's not advised, only as a last-minute resort). Even though Flutter advises to m[ock plugins](https://docs.flutter.dev/testing/testing-plugins) to test them, I can't do it because it's hidden under `flutter-quill`'s code.

I've only achieved 65% code coverage but this blocker doesn't allow me to go further.

Would it make sense to finish the README to have this working and give `visual-editor` a try in this same repo (although on a different PR)? Testing may be different and easier, as `visual-editor` is purportedly much better documented, contrary to `flutter-quill` which lacks tests.

nelsonic commented:
@LuchoTurtle yeah, 65% is quite low. Perhaps it's worth having a small paragraph in the `README.md` explaining what _can't_ be tested with `flutter-quill` ... 

Images should _always_ be uploaded to the "3rd" party service and then included via `URL`.

LuchoTurtle commented:
I've added the rest of the documentation and added all of the details into the `README` file.
After this, I've tried going through platform channels in a last attempt to override the behaviour from `image-picker` inside `flutter-quill` by using [`setMockMethodCallHandler`](https://stackoverflow.com/questions/52028969/testing-flutter-code-that-uses-a-plugin-and-platform-channel) and *although I got some progress by running the mocked overridden version*, I keep getting this following error:

```
The following PlatformException was thrown running a test:
PlatformException(error, Invalid argument: Instance of 'XFile', null, null)

When the exception was thrown, this was the stack:
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:652:7)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:310:18)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)
```

I feel like I've spent too much already on trying to get this to work, so I wager it makes sense to tackle uploading to `S3` and rendering on the web properly.

LuchoTurtle commented:
Following https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/issues/1#issuecomment-1707962562, I'm submitting this for review.

It does not *yet* upload images to `S3`, as I believe this should be contained on a different PR and because this is already a big one to review.

nelsonic commented on README.md:
```diff
@@ -1,2 +1,1655 @@
+This means everything is correctly set up!
+We are ready to start implementing!
+
+### **Make sure your `flutter` is up-to-date!
```
```suggestion
### Make sure your `flutter` is up-to-date!
```

nelsonic commented:
Just ran `flutter test --coverage` on this branch on my `localhost` got:

<img width="949" alt="image" src="https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/assets/194400/99e9c6b4-dd88-4726-bc3e-f5123974679b">

00:04 +0: loading /Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart                                                ../../.pub-cache/hosted/pub.dev/flutter_quill-7.4.1/lib/src/widgets/raw_editor.dart:353:7: Error: No named parameter with the name
'onLiveTextInput'.
      onLiveTextInput: null,
      ^^^^^^^^^^^^^^^
/opt/homebrew/Caskroom/flutter/3.3.8/flutter/packages/flutter/lib/src/widgets/editable_text.dart:1832:38: Context: Found this
candidate, but the arguments don't match.
  static List<ContextMenuButtonItem> getEditableButtonItems({
                                     ^^^^^^^^^^^^^^^^^^^^^^
00:08 +0 -1: loading /Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart [E]
  Failed to load "/Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart": Compilation failed for testPath=/Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart

To run this test again: /opt/homebrew/Caskroom/flutter/3.3.8/flutter/bin/cache/dart-sdk/bin/dart test /Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart -p vm --plain-name 'loading /Users/n/code/flutter-wysiwyg-editor-tutorial/test/widget_test.dart'
00:08 +0 -1: Some tests failed.

```sh
flutter doctor -v

[✓] Flutter (Channel stable, 3.10.6, on macOS 13.5.1 22G90 darwin-arm64, locale en-GB)
    • Flutter version 3.10.6 on channel stable at /opt/homebrew/Caskroom/flutter/3.3.8/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f468f3366c (9 weeks ago), 2023-07-12 15:19:05 -0700
    • Engine revision cdbeda788a
    • Dart version 3.0.6
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/n/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /Users/n/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.82.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.72.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.5.1 22G90 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 116.0.5845.187
    ! Error: Failed to prepare the device for development. Please unlock and reconnect the device. (code 806)

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

@LuchoTurtle please check. 🙏

nelsonic commented:
Upgrading `Flutter` on my `localhost`:
```sh
flutter upgrade --force
Upgrading Flutter to 3.13.4 from 3.10.6 in /opt/homebrew/Caskroom/flutter/3.3.8/flutter...
```

To see if that's the problem. ⏳

nelsonic commented:
```sh
Flutter 3.13.4 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 367f9ea16b (3 days ago) • 2023-09-12 23:27:53 -0500
Engine • revision 9064459a8b
Tools • Dart 3.1.2 • DevTools 2.25.0

Running flutter doctor...
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.13.4, on macOS 13.5.1 22G90 darwin-arm64, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.3)
[✓] VS Code (version 1.82.0)
[✓] Connected device (2 available)
    ! Error: Failed to prepare the device for development. Please unlock and reconnect the device. (code 806)
[✓] Network resources

• No issues found!
```

Then:
```
flutter test
```

Output:

```sh
...
> vm_service 11.7.1 (was 11.3.0) (11.10.0 available)
+ web 0.1.4-beta
  win32 5.0.6 (5.0.7 available)
  xdg_directories 1.0.2 (1.0.3 available)
  xml 6.3.0 (6.4.2 available)
Changed 8 dependencies!
00:11 +3: All tests passed!
```

All good. ✅ 

<img width="850" alt="image" src="https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/assets/194400/4431abfe-f80c-4f56-a3c8-9d1013e495cd">

🙄

nelsonic commented:
Is it not _possible_ to test this file? 💭 
<img width="1159" alt="image" src="https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/assets/194400/744e000b-7288-42b3-98aa-89cc1839b395">

nelsonic commented:
Converted your demo video to `GIF` and double-speed: https://ezgif.com/optimize/ezgif-2-26cec1fc55.gif

![dwyl-wysiwyg-demo-optimised](https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/assets/194400/08c6daff-33be-4f1a-9077-012cfd52b72c)

nelsonic commented:
Image uploads don't work on my `localhost` in `Chrome`:
<img width="825" alt="image" src="https://github.com/dwyl/flutter-wysiwyg-editor-tutorial/assets/194400/fc2d615d-be0d-468f-9dd1-670f0adf922e">

This is "OK" as it will be fixed by using `visual-editor` in the follow-up. ✅

nelsonic reviewed (approved):
Thanks @LuchoTurtle :shipit:
